// @remotion/tailwind-v4/dist/cjs/enable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.enableTailwind = void 0;
/**
 * @description A function that modifies the default Webpack configuration to make the necessary changes to support Tailwind.
 * @see [Documentation](https://www.remotion.dev/docs/tailwind-v4/enable-tailwind)
 */
const enableTailwind = (currentConfiguration) => {
    var _a;
    return {
        ...currentConfiguration,
        module: {
            ...currentConfiguration.module,
            rules: [
                ...(((_a = currentConfiguration.module) === null || _a === void 0 ? void 0 : _a.rules)
                    ? currentConfiguration.module.rules
                    : []).filter((rule) => {
                    var _a;
                    return rule && rule !== '...' && !((_a = rule.test) === null || _a === void 0 ? void 0 : _a.toString().includes('.css'));
                }),
                {
                    test: /\.css$/i,
                    use: [
                        require.resolve('style-loader'),
                        require.resolve('@remotion/bundler/css-loader'),
                        require.resolve('@tailwindcss/webpack'),
                    ],
                },
            ],
        },
    };
};
exports.enableTailwind = enableTailwind;
